app/src/main/java/com/sanjeevyadavit/magecart/viewmodel/SignupViewModel.kt

```diff
@@ -0,0 +1,71 @@
+package com.sanjeevyadavit.magecart.viewmodel
+
+import android.app.Application
+import android.util.Log
+import android.widget.Toast
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.viewModelScope
+import com.sanjeevyadavit.magecart.model.Customer
+import com.sanjeevyadavit.magecart.model.LoginBodyRequest
+import com.sanjeevyadavit.magecart.model.SignupBodyRequest
+import com.sanjeevyadavit.magecart.model.SignupResponse
+import com.sanjeevyadavit.magecart.service.ApiInterface
+import com.sanjeevyadavit.magecart.service.RetrofitClient
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import retrofit2.Call
+import retrofit2.Response
+
+class SignupViewModel(app: Application): AndroidViewModel(app) {
+    val email: MutableLiveData<String> = MutableLiveData("")
+    val password: MutableLiveData<String> = MutableLiveData("")
+    val firstName: MutableLiveData<String> = MutableLiveData("")
+    val lastName: MutableLiveData<String> = MutableLiveData("")
+
+    fun signup(){
+        if(handleEmptyValues()) return
+        makeSignupApiCall()
+    }
+
+    private fun handleEmptyValues(): Boolean{
+        if(email.value!!.isEmpty()  || password.value!!.isEmpty() || firstName.value!!.isEmpty() || lastName.value!!.isEmpty()) {
+            showToast("Fields cannot be empty")
+            return true;
+        }
+        return false
+    }
+
+    private fun makeSignupApiCall(){
+        viewModelScope.launch(Dispatchers.IO) {
+            val facade: ApiInterface = RetrofitClient.getInstance().create(ApiInterface::class.java)
+            facade.signup(SignupBodyRequest(Customer(email.value!!, firstname = firstName.value!!, lastname = lastName.value!!), password.value!!)).enqueue(
```

Comment:
same.. no `!!` and use a class that gives you singleton instance of ApiInterface, like the way done for Retrofit.